File: label-aggregator-springboot3/src/main/java/com/example/aggregator/LabelClient.java
package com.example.aggregator;

import org.springframework.aot.hint.annotation.RegisterReflectionForBinding;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;

import reactor.core.publisher.Mono;

@Service
public class LabelClient {
    private final WebClient client = WebClient.create("http://localhost:8090");

    @RegisterReflectionForBinding(LabelResult.class)
    public Mono<LabelResult> getAsync(String key) {
        return client.get()
                .uri("/api/" + key)
                .retrieve()
                .bodyToMono(LabelResult.class);
    }

    

}

File: label-aggregator-springboot3-kotlin/src/main/kotlin/com/example/aggregator/LabelClient.kt
package com.example.aggregator

import kotlinx.coroutines.*
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.awaitBody
import org.springframework.web.reactive.function.client.awaitExchange

@Service
class LabelClient(private val client: WebClient) {
    @Value("\${label-client.url}") private lateinit var baseURL: String

    suspend fun getAsync(key: String): LabelResult =
            client.get().uri("$baseURL/${key}").awaitExchange { clientResponse ->
                clientResponse.awaitBody<LabelResult>()
            }

    data class LabelResult(val key: String, val label: String)
}
